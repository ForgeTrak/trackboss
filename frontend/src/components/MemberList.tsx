import { useDisclosure } from '@chakra-ui/react';
import _ from 'lodash';
import React, { useContext, useEffect, useState } from 'react';
import DataTable from 'react-data-table-component';
import { ErrorResponse } from '../../../src/typedefs/errorResponse';
import { Member } from '../../../src/typedefs/member';
import { UserContext } from '../contexts/UserContext';
import { getAllBoardMembersForCurrentYear } from '../controller/boardMember';
import { getMemberList } from '../controller/member';
import MemberSummaryModal from './MemberSummaryModal';

const columns: any = [
    {
        name: 'Last Name',
        selector: (row: Member) => row.lastName,
        sortable: true,
    },
    {
        name: 'First Name',
        selector: (row: Member) => row.firstName,
        sortable: true,
    },
    {
        name: 'Membership Type',
        selector: (row: Member) => row.membershipType,
        sortable: true,
    },
    {
        name: 'Role',
        selector: (row: Member) => row.memberType,
        sortable: true,
    },
];

const customStyles = {
    rows: {
        style: {
            minHeight: '65px',
        },
    },
    headCells: {
        style: {
            marginTop: '90px',
            paddingTop: '0',
            fontSize: '1.5em',
            backgroundColor: '#f9f9f9',
            color: '#626262',
        },
    },
    cells: {
        style: {
            fontSize: '1.2em',
        },
    },
};

export default function MemberList() {
    const [selectedMember, setSelectedMember] = useState<Member>();
    const [cells, setCells] = useState<Member[]>([]);
    const { state } = useContext(UserContext);
    const [error, setError] = useState<ErrorResponse | undefined>(undefined);
    const [dirty, setDirty] = useState<boolean>(false);
    const { onClose, isOpen, onOpen } = useDisclosure({ onClose: () => setDirty((oldDirty) => !oldDirty) });

    useEffect(() => {
        async function getData() {
            const c: Member[] | ErrorResponse = await getMemberList(state.token);
            const boardMembers = await getAllBoardMembersForCurrentYear(state.token);
            if ('reason' in c) {
                setError(c);
            } else {
                if ('reason' in boardMembers) {
                    // do nothing, an error here isn't a big deal, it just means we can't show baord member data
                } else {
                    _.forEach(boardMembers, (boardMember) => {
                        _.forEach(c, (member: Member) => {
                            if (member.memberId === boardMember.memberId) {
                                member.memberType = `Board Member - ${boardMember.title}`;
                                member.boardMemberData = boardMember;
                                return false;
                            }
                            return true;
                        });
                    });
                }
                setCells(_.filter(c, (member) => member.active));
                setError(undefined);
            }
        }
        getData();
    }, [dirty]);
    if (error) {
        return (
            <div>
                {error.reason}
            </div>
        );
    }
    return (
        <div>
            <DataTable
                columns={columns}
                data={cells}
                fixedHeaderScrollHeight="300px"
                highlightOnHover
                pagination
                responsive
                subHeaderWrap
                defaultSortFieldId={1}
                customStyles={customStyles}
                paginationComponentOptions={
                    {
                        selectAllRowsItem: true,
                    }
                }
                onRowClicked={
                    (row: Member) => {
                        setSelectedMember(row);
                        onOpen();
                    }
                }
            />
            {
                selectedMember && (
                    <MemberSummaryModal
                        isOpen={isOpen}
                        onClose={onClose}
                        memberInfo={selectedMember}
                    />
                )
            }
        </div>
    );
}
