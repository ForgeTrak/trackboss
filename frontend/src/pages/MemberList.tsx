import React from 'react';
import { ChakraProvider } from '@chakra-ui/react';
import theme from '../theme';
import Header from '../components/Header';
import MemberSummaryModal from '../components/MemberSummaryModal';
import { Member } from '../../../src/typedefs/member';
import { Bike } from '../../../src/typedefs/bike';

const member: Member = {
    memberId: 1,
    membershipAdmin: 'true',
    active: true,
    memberType: 'board',
    firstName: 'test',
    lastName: 'member',
    phoneNumber: '1234',
    email: 'user@example.com',
    uuid: '',
    occupation: 'secretary',
    birthdate: '',
    dateJoined: '',
    address: '',
    city: '',
    state: '',
    zip: '',
    lastModifiedDate: '',
    lastModifiedBy: '',
};

const memberFamily: Member[] = [
    {
        memberId: 1,
        membershipAdmin: 'true',
        active: true,
        memberType: 'member',
        firstName: 'test',
        lastName: 'member',
        phoneNumber: '1234',
        email: 'user@example.com',
        uuid: '',
        occupation: '',
        birthdate: '',
        dateJoined: '',
        address: '',
        city: '',
        state: '',
        zip: '',
        lastModifiedDate: '',
        lastModifiedBy: '',
    },
    {
        memberId: 2,
        membershipAdmin: 'string',
        active: true,
        memberType: 'member',
        firstName: 'Jon',
        lastName: 'Smith',
        phoneNumber: '1234',
        email: 'user@example.com',
        uuid: '',
        occupation: '',
        birthdate: '',
        dateJoined: '',
        address: '',
        city: '',
        state: '',
        zip: '',
        lastModifiedDate: '',
        lastModifiedBy: '',
    },
];

const memberBikes: Bike[] = [
    {
        bikeId: 0,
        year: '2012',
        make: 'honda',
        model: 'shitbox',
        membershipAdmin: 'string',
    },
];

function MemberList() {
    return (
        <ChakraProvider theme={theme}>
            <Header title="Members" activeButtonId={3} />
            <MemberSummaryModal memberInfo={member} memberFamily={memberFamily} memberBikes={memberBikes} />
        </ChakraProvider>
    );
}

export default MemberList;
