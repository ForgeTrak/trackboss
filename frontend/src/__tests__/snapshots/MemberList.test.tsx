import React from 'react';
import renderer, { act } from 'react-test-renderer';
import MemberList from '../../components/MemberList';

describe('Member List component', () => {
    it('renders correctly', () => {
        let list;
        act(() => {
            list = renderer.create(<MemberList />);
        });
        expect(list.toJSON()).toMatchSnapshot();
    });
});
