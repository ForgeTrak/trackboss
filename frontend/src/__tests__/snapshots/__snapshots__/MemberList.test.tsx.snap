// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Member List component renders correctly 1`] = `
<DocumentFragment>
  <div>
    <div
      class="sc-dlVxhl fSQziN"
    >
      <div
        class="sc-fKVqWL XlRpR"
      >
        <div
          class="sc-bdvvtL hLGqlg rdt_Table"
          role="table"
        >
          <div
            class="sc-gsDKAQ fbhklE rdt_TableHead"
            role="rowgroup"
          >
            <div
              class="sc-dkPtRN hCBnvI rdt_TableHeadRow"
              role="row"
            >
              <div
                class="sc-hKwDye sc-eCImPb sc-iqseJM crvHGE fvAeSn iwmeYY rdt_TableCol"
                data-column-id="1"
              >
                <div
                  class="sc-crHmcD fFBlx rdt_TableCol_Sortable"
                  data-column-id="1"
                  data-sort-id="1"
                  role="columnheader"
                  tabindex="0"
                >
                  <div
                    class="sc-egiyK lnndaO"
                    data-column-id="1"
                  >
                    Name
                  </div>
                  <span
                    class="sc-kDTinF gqExIP"
                  >
                    ▲
                  </span>
                </div>
              </div>
              <div
                class="sc-hKwDye sc-eCImPb sc-iqseJM crvHGE fvAeSn iwmeYY rdt_TableCol"
                data-column-id="2"
              >
                <div
                  class="sc-crHmcD fFBlx rdt_TableCol_Sortable"
                  data-column-id="2"
                  data-sort-id="2"
                  role="columnheader"
                  tabindex="0"
                >
                  <div
                    class="sc-egiyK lnndaO"
                    data-column-id="2"
                  >
                    Role
                  </div>
                  <span
                    class="sc-kDTinF gqExIP"
                  >
                    ▲
                  </span>
                </div>
              </div>
            </div>
          </div>
          <div
            class="sc-hGPBjI fuBVgf rdt_TableBody"
            role="rowgroup"
          >
            <div
              class="sc-jrQzAO kCxImp rdt_TableRow"
              id="row-0"
              role="row"
            >
              <div
                class="sc-hKwDye sc-eCImPb sc-jRQBWg fDgsXD fvAeSn bsgpsK rdt_TableCell"
                data-column-id="1"
                data-tag="allowRowEvents"
                id="cell-1-undefined"
                role="gridcell"
              >
                <div
                  data-tag="allowRowEvents"
                >
                  Squeak Trainywhel
                </div>
              </div>
              <div
                class="sc-hKwDye sc-eCImPb sc-jRQBWg fDgsXD fvAeSn bsgpsK rdt_TableCell"
                data-column-id="2"
                data-tag="allowRowEvents"
                id="cell-2-undefined"
                role="gridcell"
              >
                <div
                  data-tag="allowRowEvents"
                >
                  Admin
                </div>
              </div>
            </div>
            <div
              class="sc-jrQzAO kCxImp rdt_TableRow"
              id="row-1"
              role="row"
            >
              <div
                class="sc-hKwDye sc-eCImPb sc-jRQBWg fDgsXD fvAeSn bsgpsK rdt_TableCell"
                data-column-id="1"
                data-tag="allowRowEvents"
                id="cell-1-undefined"
                role="gridcell"
              >
                <div
                  data-tag="allowRowEvents"
                >
                  Isobel Jennery
                </div>
              </div>
              <div
                class="sc-hKwDye sc-eCImPb sc-jRQBWg fDgsXD fvAeSn bsgpsK rdt_TableCell"
                data-column-id="2"
                data-tag="allowRowEvents"
                id="cell-2-undefined"
                role="gridcell"
              >
                <div
                  data-tag="allowRowEvents"
                >
                  Membership Admin
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div>
      <nav
        class="sc-iJKOTD bhoXfZ rdt_Pagination"
      >
        <span
          class="sc-bYoBSM sc-ikJyIC fXzwgx bWTWQu"
        >
          Rows per page:
        </span>
        <div
          class="sc-llYSUQ fQmYxz"
        >
          <select
            aria-label="Rows per page:"
            class="sc-cxpSdN fWWeYy"
          >
            <option
              selected=""
              value="10"
            >
              10
            </option>
            <option
              value="15"
            >
              15
            </option>
            <option
              value="20"
            >
              20
            </option>
            <option
              value="25"
            >
              25
            </option>
            <option
              value="30"
            >
              30
            </option>
          </select>
          <svg
            height="24"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M7 10l5 5 5-5z"
            />
            <path
              d="M0 0h24v24H0z"
              fill="none"
            />
          </svg>
        </div>
        <span
          class="sc-bYoBSM sc-kLwhqv fXzwgx emVwZU"
        >
          1-2 of 2
        </span>
        <div
          class="sc-ezbkAF kNbUsB"
        >
          <button
            aria-disabled="true"
            aria-label="First Page"
            class="sc-giYglK gPLhoV"
            disabled=""
            id="pagination-first-page"
            type="button"
          >
            <svg
              aria-hidden="true"
              height="24"
              role="presentation"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M18.41 16.59L13.82 12l4.59-4.59L17 6l-6 6 6 6zM6 6h2v12H6z"
              />
              <path
                d="M24 24H0V0h24v24z"
                fill="none"
              />
            </svg>
          </button>
          <button
            aria-disabled="true"
            aria-label="Previous Page"
            class="sc-giYglK gPLhoV"
            disabled=""
            id="pagination-previous-page"
            type="button"
          >
            <svg
              aria-hidden="true"
              height="24"
              role="presentation"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M15.41 7.41L14 6l-6 6 6 6 1.41-1.41L10.83 12z"
              />
              <path
                d="M0 0h24v24H0z"
                fill="none"
              />
            </svg>
          </button>
          <button
            aria-disabled="true"
            aria-label="Next Page"
            class="sc-giYglK gPLhoV"
            disabled=""
            id="pagination-next-page"
            type="button"
          >
            <svg
              aria-hidden="true"
              height="24"
              role="presentation"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M10 6L8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
              />
              <path
                d="M0 0h24v24H0z"
                fill="none"
              />
            </svg>
          </button>
          <button
            aria-disabled="true"
            aria-label="Last Page"
            class="sc-giYglK gPLhoV"
            disabled=""
            id="pagination-last-page"
            type="button"
          >
            <svg
              aria-hidden="true"
              height="24"
              role="presentation"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M5.59 7.41L10.18 12l-4.59 4.59L7 18l6-6-6-6zM16 6h2v12h-2z"
              />
              <path
                d="M0 0h24v24H0V0z"
                fill="none"
              />
            </svg>
          </button>
        </div>
      </nav>
    </div>
  </div>
</DocumentFragment>
`;
